src/Featurizers/Components/WindowFeaturizerBase.h

```diff
@@ -0,0 +1,244 @@
+                !((_cur_increment == _max_increments) ^ (rhs._cur_increment == rhs._max_increments));
+        }
+
+        template<class OtherType>
+        bool operator != (const CircularIterator<OtherType>& rhs) const {
+            return !operator==(rhs);
+        }
+
+        T& operator* () const {
+            assert(_cur_increment != _max_increments);
+            return *(_itr+_cur_index);
+        }
+
+        T* operator-> () const {
+            assert(_cur_increment != _max_increments);
+            return (_itr+_cur_index);
+        }
+
+        // One way conversion: iterator -> const_iterator
+        operator CircularIterator<const T>() const {
+            return CircularIterator<const T>(_itr);
+        }
+
+    };
+
+    /////////////////////////////////////////////////////////////////////////
+    ///  \class         CircularBuffer
+    ///  \brief         A custom container class created for operations with shifted windows
+    ///                 The goal is to minimize the memory allocation so when the item limit
+    ///                 has been reached it overwrites the oldest item it is storing without
+    ///                 allocating new memory.
+    ///
+    template <class T>
+    class CircularBuffer {
+
+        // Maximum number of elements this container can hold. When this number is reached
+        // the oldest value is overwritten.
+        size_t                                  _max_size;
+
+        // The start position of this circular buffer. This class is implemented by wrapping around
+        // a vector so the "start" will change based on how much data has been overwritten.
+        size_t                                  _start_offset;
+
+        // Current number of element held by this container. It start from 0 and grows to _max_size
+        size_t                                  _cur_size;
+
+        // The vector that holds the actual data. This is set to hold _max_size elements and then
+        // is never resized.
+        std::vector<T>                          _data;
+
+    public:
+        // ----------------------------------------------------------------------
+        // |  Public Data
+        typedef CircularIterator<T>             iterator;
+        typedef CircularIterator<const T>       const_iterator;
+
+        // ----------------------------------------------------------------------
+        // |  Public Methods
+        CircularBuffer(size_t max_size) : _max_size(max_size), _start_offset(0),  _cur_size(0) {
+            _data.reserve(_max_size);
+        }
+
+        iterator begin() {
+            // When _cur_size is 0 we want to return an "end" iterator.
+            if (_cur_size == 0) {
+                return iterator();
+            }
+            return iterator(&_data[0], _max_size, _cur_size, _start_offset);
+        }
+
+        const_iterator begin() const {
+            return begin();
```

Comment:
>return begin(); [](start = 12, length = 15)

I'm not sure how this code is working. The type CircularIterator<T> should not be convertible to CircularIterator<T const>.

Reply:
Do you mean convert to const CircularIterator? There's a method in the CircularIterator class that can do the conversion.

Reply:
Made the changes discussed offline. Removed the conversion code and added `_impl` methods.